httpie/output/lexers/json.py

```diff
@@ -0,0 +1,44 @@
+from pygments.token import Token
+from pygments.lexers.data import JsonLexer
+
+
+class EnhancedJsonLexer(JsonLexer):
+    """
+    Enhanced JSON lexer for Pygments.
+
+    It adds support for eventual data prefixing the actual JSON body.
+
+    """
+
+    body_prefix_sep = '@'
+    body_prefix_token = Token.Error
+
+    # Treat `body_prefix_sep` as a punctuation character.
+    punctuations = {*JsonLexer.punctuations, body_prefix_sep}
+
+    def get_tokens_unprocessed(self, text):
+        """Parse JSON data."""
+
+        in_body_prefix = False
+
+        for index, token, value in super().get_tokens_unprocessed(text):
+            if token is Token.Punctuation and self.body_prefix_sep in value:
+                # `value` can be a single character or a sequence of punctuation characters.
+                # We need to iterate over the whole `value` to properly handle the body prefix
+                # separator from actual JSON punctuation characters.
+                for char in value:
+                    if char == self.body_prefix_sep:
+                        # Body prefix.
+                        in_body_prefix = not in_body_prefix
+                    elif in_body_prefix:
+                        # Punctuation character part of the body prefix.
+                        yield index, self.body_prefix_token, char
+                    else:
+                        # Punctuation character outside the body prefix.
+                        yield index, token, char
+            elif in_body_prefix:
+                # `value` is part of the body prefix.
+                yield index, self.body_prefix_token, value
+            else:
+                # Outside the body prefix.
+                yield index, token, value
```

Comment:
You can just have a simple lexer that delegates to the JSON one anything after the prefix (needs some testing). No additional treatment for the prefix is needed when it comes to colors, so you can simplify `ColorFormatter` again.

```python
from pygments.lexer import using, RegexLexer

class EnhancedJsonLexer(RegexLexer):
    flags = re.IGNORECASE | re.DOTALL
    tokens = {
        'root': [
            (r'[^{\[]+', Token.Error),  # Prefix
            (r'.+', using(JsonLexer)),
        ],
    }
```

Reply:
Much simpler :+1:

Reply:
My brain kept thinking about it in the background and the above probably breaks with naked strings, numbers, and booleans (we should have tests for those). A more explicit handling of the prefix should be more robust, something like:


```python
import re
from pygments.lexer import using, RegexLexer, bygroups


PrefixToken = Token.Error

class EnhancedJsonLexer(RegexLexer):
    flags = re.IGNORECASE | re.DOTALL
    tokens = {
        'root': [
            (r'([^{\[]+)([{\]].+)', bygroups(PrefixToken, using(JsonLexer))),
            (r'.+', using(JsonLexer)),
        ],
    }
```